spec/reactors/consensus.md

```diff
@@ -0,0 +1,261 @@
+
+For every block proposal, peers start gossiping their obtained proposal and constituent block parts to their connected peers.
+Both sending and receiving ends monitor the block parts they've exchanged with their counterpart (either dispatched to or received from) and mark it as seen by the other peer.
+Peers keep exchanging block parts of a proposal until either 1) all block parts of the intended proposal have been successfully transmitted between the two peers or 2) one of the peer's round state updates (and points to a new height and round with a different proposal).
+In the latter case, the peer whose state has advanced still sends block parts to the other peer until all the parts are transmitted or until the receiving peer's round state is also updated.
+
+Worst Case Scenario: The worst case occurs when both peers coincidentally choose the same block part index at the same moment and initiate the sending process concurrently, a scenario that's rather unlikely.
+The outcome of this is that the cumulative number of block parts transmitted between the two peers (sent and received) equals `2 * proposal(H,R).total`.
+Likewise, 2 instances of the proposal header are also transmitted between the peers.
+
+Best Case Scenario: Ideally, only one instance of each block part is exchanged between the two peers (the opposite of the worst case).
+Consequently, the aggregate number of block parts transferred (both sent and received) between the peers is `proposal(H,R).total`. 
+Also, only one instance of the proposal header is transmitted between the two peers.
+This number can further reduce if peers acquires block parts from additional connections, thereby advancing to the subsequent height, round, or proposal.
+In this scenario, both parties notify one another that they have advanced to the next height and round, hence they both stop transmitting block parts of that proposal to each other.
+
+Based on above, it can be established that one network health indicator is that the cumulative number of block parts sent and received over each p2p connection over `DataChannel` should not surpass the total block parts specified in the proposal for a particular height and round.
+This should hold true even when one of the  two ends of communication lags behind and is catching up by obtaining block parts of the past blocks. 
+
+[//]: # (TODO: will verify this by inspecting the Prometheus metrics for the number of block parts sent and received over each p2p connection. Alternatively, will develop a go test to verify this. -->)
+
+### Questions and Optimization Ideas
+
+1. In the event of a connection disruption, is the peer's round state reset, or does the process pick up from the last known point?
+1. [Optimization idea] could other peers halt the transmission of block parts to a peer that reaches the prevote step (taking prevote step as an indication that the node must have possessed the entire block)?
```

Comment:
I'm not sure if we want to list the [#904](https://github.com/cometbft/cometbft/pull/904) optimization here as that's similar, however it is breaking. This suggestion seems like it is not breaking.

Reply:
Right, the one in the doc would not be a breaking change. I am also going to include the https://github.com/cometbft/cometbft/pull/904 in the doc to keep it in our radar.

Reply:
Done https://github.com/celestiaorg/celestia-core/pull/1129/commits/a75469838545e496c3b6dd121ce058e8946b1e7d

Reply:
Tracking issue for new optimization idea https://github.com/celestiaorg/celestia-core/issues/1134